mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp

```diff
@@ -1551,8 +1551,21 @@ LogicalResult ReturnOp::verify() {
 //===----------------------------------------------------------------------===//
 
 LogicalResult ResumeOp::verify() {
-  if (!getValue().getDefiningOp<LandingpadOp>())
-    return emitOpError("expects landingpad value as operand");
+  auto parentFunc = getOperation()->getParentOfType<FunctionOpInterface>();
+  assert(parentFunc && "Expecting parent function");
+  const auto ty = getOperand().getType();
+  if (!parentFunc
+           ->walk([ty](LandingpadOp landingpad) {
+             return landingpad.getType() == ty
+                        ? WalkResult::interrupt() // Just an operation needed
```

Comment:
> The ‘resume’ instruction requires one argument, which must have the same type as the result of any ‘landingpad’ instruction in the same function.

It sounds to me that the result type of any landingpad not the same as the resume operand type is not allowed.
Did you check that LLVM only requires one landingpad with the same type?

Reply:
That's not allowed because you cannot have `landingpad` operations with different types in the same function.

LLVM's verifier is order-dependent, so it takes the result type of the first `landingpad` or `resume` instruction as the "true" type. If any of the following operations do not match this type, it'll cause an error.

If we want to mimic this, we'd have to walk every operation till we find a `langingpad` or `resume` and take that as the true value. Our check here perfectly mimics the spec, whereas LLVM's does not (they'd allow something like):

```llvm
declare i32 @foo(i32)

define i32 @caller(i32 %arg0) personality i32 (i32)* @foo {
entry:
  resume i32 1
}
```

which clearly violates the spec.

I suggest we implement a verifier matching the spec, but I'm open to changes here if we agree.